client/ayon_nuke/api/plugin.py

```diff
@@ -309,6 +310,39 @@ def set_selected_nodes(self, pre_create_data):
         else:
             self.selected_node = None
 
+    def update_instances(self, update_list):
+        for created_inst, changes in update_list:
+            instance_node = created_inst.transient_data["node"]
+
+            # update instance node name if product name changed
+            if "productName" in changes.changed_keys:
+                changed_data = {
+                    "productName": changes["productName"].new_value or (
+                        created_inst["productName"]),
+                    "folderPath": changes["folderPath"].new_value or (
+                        created_inst["folderPath"]
+                    ),
+                    "task": changes["task"].new_value or (
+                        created_inst["task"]
+                    ),
+                    "productType": changes["productType"].new_value or (
+                        created_inst["productType"]
+                    )
```

Comment:
This `changed_data` dict isn't actually the changed data.
This `x` or `y` logic doesn't need to be there if you just call: 

```python
data = created_inst.data_to_store()
keys = ("productName", "folderPath", "task", "productType")
context_values = {key: data[key] for key in keys}
reset_write_node_filepath(instance_node, changed_data)
```

However, I'd argue that even better would be if the `reset` logic was on the Creator and you'd just pass it the created instance so that the function call simplifies.

```python
def update_write_node_filepath(self, created_inst, changes):
    keys = ("productName", "folderPath", "task", "productType")
    if not any(key in changes for key in keys):
        # No relevant changes, no need to update
        return

    # Update values with new formatted path
    instance_node = created_inst.transient_data["node"]
    formatting_data = copy.deepcopy(data)
    write_node = nuke.allNodes(group=instance_node, filter="Write")[0]
    formatting_data.update({
        "fpath_template": (
        "{work}/renders/nuke/{subset}/{subset}.{frame}.{ext}"),
        "ext": write_node["file_type"].value()
    })
    anatomy_filled = format_anatomy(formatting_data)

    # build file path to workfiles
    fdir = str(
        anatomy_filled["work"]["default"]["directory"]
    ).replace("\\", "/")
    formatting_data["work"] = fdir
    fpath = StringTemplate(formatting_data["fpath_template"]).format_strict(
        formatting_data)
    write_node["file"].setValue(fpath)
    instance_node["name"].setValue(formatting_data["productName"])
```

The issue with the function in `lib` now is that it expects `data` but it is not clear, for a library function, what that data should need. It may be better to then turn it into passing the product name, etc. as separate arguments - but I don't think there's huge benefit now in moving it to `lib`.